File: baekjoon/src/binarySearch/No2805.java
package binarySearch;

import java.io.*;
import java.util.StringTokenizer;

/**
 * https://www.acmicpc.net/problem/2805
 */
public class No2805 {
    public static final BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
    public static final BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
    public static StringTokenizer tokenizer;

    public static int numberOfTrees;
    public static int requiredLength;
    public static int[] trees;
    public static int high = Integer.MAX_VALUE;
    public static int low = 0;

    public static void setValues() throws IOException {
        tokenizer = new StringTokenizer(reader.readLine());
        numberOfTrees = Integer.parseInt(tokenizer.nextToken());
        requiredLength = Integer.parseInt(tokenizer.nextToken());

        tokenizer = new StringTokenizer(reader.readLine());
        trees = new int[numberOfTrees];
        for (int i = 0; i < numberOfTrees; i++)
            trees[i] = Integer.parseInt(tokenizer.nextToken());
    }

    public static int getMaxHeight() {
        int mid = (high + low) / 2;
        while (low + 1 < high) {
            if (isValid(mid))
                low = mid;
            else
                high = mid;
            mid = (high + low) / 2;
        }
        return mid;
    }

    private static boolean isValid(int mid) {
        long sum = 0;
        for (int i = 0; i < numberOfTrees; i++) {
            if (trees[i] > mid)
                sum += trees[i] - mid;
        }
        return sum >= requiredLength;
    }

    public static void main(String[] args) throws IOException {
        setValues();
        writer.write(String.valueOf(getMaxHeight()));
        writer.flush();
    }
}

File: yourssu/src/week2/No2805.kt
package week2

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

/**
 * https://www.acmicpc.net/problem/2805
 */
private val reader = BufferedReader(InputStreamReader(System.`in`))
private val writer = BufferedWriter(OutputStreamWriter(System.out))

fun main() {
    val requiredHeight = reader.readLine().split(" ").map { it.toLong() }[1]
    val heights = reader.readLine().split(" ").map { it.toLong() }
    writer.write("${getHeight(heights, requiredHeight)}")
    writer.flush()
}

private fun getHeight(heights: List<Long>, requiredHeight: Long): Int {
    var low = 0
    var high = 1000000000 + 1

    while (low + 1 < high) {
        val height = (low + high) / 2
        if (heights.sumOf { (it - height).coerceAtLeast(0) } < requiredHeight)
            high = height
        else
            low = height
    }

    return low
}